File: src/com/leetcode/Medium/ProductOfArrayExceptSelf.java
package com.leetcode.Medium;

// 238. Product of Array Except Self
public class ProductOfArrayExceptSelf {

    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;

        int[] L = new int[n];
        int[] R = new int[n];
        int[] arr = new int[n];

        L[0] = 1;
        for (int i = 1; i < n; i++) {
            L[i] = nums[i - 1] * L[i - 1];
        }

        R[n - 1] = 1;
        for (int i = n - 2; i >= 0; i--) {
            R[i] = nums[i + 1] * R[i + 1];
        }

        for (int i = 0; i < n; i++) {
            arr[i] = L[i] * R[i];
        }
        return arr;
    }
}

File: JavaScript/Medium/ProductOfArrayExceptSelf.js
/**
 * 238. Product of Array Except Self
 * @param {number[]} nums
 * @return {number[]}
 */
const productExceptSelf = function (nums) {
  let len = nums.length;
  let left = new Array(len), right = new Array(len);
  left[0] = 1, right[len - 1] = 1;

  for (let i = 1; i < len; i++) {
    left[i] = left[i - 1] * nums[i - 1];
  }

  for (let i = len - 2; i >= 0; i--) {
    right[i] = right[i + 1] * nums[i + 1];
  }

  let result = new Array(len);
  for (let i = 0; i < len; i++) {
    result[i] = left[i] * right[i];
  }

  return result;
};